### PR TITLE
Implement cost integration via IRK for Convex-over-nonlinear cost 

### PR DESCRIPTION
**Pull request by FreyJo**
-  Allows for Generalized GNRK integrator (GGNRK), eXtended GNRK integrator (XGNRK)

### CODE REVIEW

**Review by sandmaennchen — COMMENTED** [review]
looks good, but I think we need an additional check in the interface regarding algebraic variables. Do we only neglect their hessian contribution or is the gradient incorrect too?

File: interfaces/acados_template/acados_template/acados_ocp_solver.py [hunk 1]
```diff
@@ -270,8 +270,9 @@ def make_ocp_dims_consistent(acados_ocp: AcadosOcp):
             "Note: There is also the option to use the external cost module with a numerical hessian approximation (see `ext_cost_num_hess`).\n"
             "OR the option to provide a symbolic custom hessian approximation (see `cost_expr_ext_cost_custom_hess`).\n")
 
-    if any([cost.cost_type_0 != "NONLINEAR_LS", cost.cost_type != "NONLINEAR_LS", cost.cost_type_e != "NONLINEAR_LS"])\
-        and opts.cost_discretization == 'INTEGRATOR':
+    supports_cost_integration = lambda type : type in ['NONLINEAR_LS', 'CONVEX_OVER_NONLINEAR']
```
[the remark this answers is not in the document]
Comment (FreyJo) [reply interfaces/acados_template/acados_template/acados_ocp_solver.py:273]: I thought it makes it easier to read.

File: acados/sim/sim_irk_integrator.c [hunk 2]
```diff
@@ -1348,14 +1382,115 @@ int sim_irk(void *config_, sim_in *in, sim_out *out, void *opts_, void *mem_, vo
...
+                    }
+
+                    // evaluate external function
+                    model->conl_cost_fun_jac_hess->evaluate(model->conl_cost_fun_jac_hess, conl_fun_jac_hess_type_in,
+                                                conl_fun_jac_hess_in, conl_fun_jac_hess_type_out, conl_fun_jac_hess_out);
+                    // hessian of outer loss function
+                    blasfeo_dpotrf_l(ny, workspace->W, 0, 0, mem->W_chol, 0, 0);
+
+                    if (nz > 0) // TODO: test this!
+                    {
```
[the remark this answers is not in the document]
Comment (FreyJo) [reply acados/sim/sim_irk_integrator.c:1448]: There is a check in Line 1060:
```
    if (opts->cost_computation)
    {
        // initialize cost_fun, cost_grad, cost_hess
        blasfeo_dvecse(nx+nu, 0.0, cost_grad, 0);
        blasfeo_dgese(nx+nu, nx+nu, 0.0, cost_hess, 0, 0);
        mem->cost_fun[0] = 0.0;
        if (nz > 0)
        {
            printf("\nIRK cost_computation not implemented for nz>0!\n\n");
            exit(1);
        }
    }
```

File: acados/sim/sim_irk_integrator.c [hunk 3]
```diff
@@ -1348,14 +1382,115 @@ int sim_irk(void *config_, sim_in *in, sim_out *out, void *opts_, void *mem_, vo
                                     1.0, cost_grad, 0, cost_grad, 0);
 
                     // cost_hess += b * tmp_nux_ny_2 * tmp_nux_ny_2^T
+                    // TODO: use syrk (exploit symmetry)
```
[the remark this answers is not in the document]
Comment (FreyJo) [reply acados/sim/sim_irk_integrator.c]: addressed in https://github.com/acados/acados/pull/976/commits/4b0f97ccf28e000a913f092f2903c9bab4563c30

**Review by sandmaennchen — APPROVED** [review]
nice!

**Pull request merged**